File: app/src/main/java/com/motorider/models/Route.java
package com.motorider.models;

import java.util.List;

public class Route {
    private String name;
    private List<Waypoint> waypoints;
    private double distance; // in kilometers
    private double duration; // in seconds
    private double curvatureScore; // 0-100 score for curvy roads
    private double elevationGain; // in meters

    public Route(String name, List<Waypoint> waypoints) {
        this.name = name;
        if (waypoints == null) {
            throw new IllegalArgumentException("Waypoints cannot be null");
        }
        this.waypoints = waypoints;
    }

    // Getters and setters
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }

    public List<Waypoint> getWaypoints() { return waypoints; }
    public void setWaypoints(List<Waypoint> waypoints) { this.waypoints = waypoints; }

    public double getDistance() { return distance; }
    public void setDistance(double distance) { this.distance = distance; }

    public double getDuration() { return duration; }
    public void setDuration(double duration) { this.duration = duration; }

    public double getCurvatureScore() { return curvatureScore; }
    public void setCurvatureScore(double curvatureScore) { this.curvatureScore = curvatureScore; }

    public double getElevationGain() { return elevationGain; }
    public void setElevationGain(double elevationGain) { this.elevationGain = elevationGain; }

    private RouteType routeType;

    public RouteType getRouteType() { return routeType; }
    public void setRouteType(RouteType routeType) { this.routeType = routeType; }
}

File: app/src/main/java/com/motorider/models/Route.kt
package com.motorider.models

import org.osmdroid.util.GeoPoint

class Route(
    val name: String,
    waypoints: List<Waypoint>
) {
    private var _waypoints: List<Waypoint> = emptyList<Waypoint>()

    init {
        require(waypoints != null) { "Waypoints cannot be null" }
        _waypoints = ArrayList(waypoints)
    }

    val waypoints: List<Waypoint>
        get() = _waypoints

    var distance: Double = 0.0
    var duration: Double = 0.0
    var curvatureScore: Double = 0.0
    var elevationGain: Double = 0.0
    var routeScore: Double = 0.0
    var routeType: RouteType? = null
    var avoidances: Set<Avoidance> = emptySet()
    var routeGeometry: List<GeoPoint>? = null
    var curvatureMetadata: CurvatureMetadata? = null

    /**
     * True when this is NOT a real route: the routing API could not be reached, so
     * distance/duration are straight-line estimates and the geometry is just the
     * waypoints joined by lines. Must be surfaced to the rider - following one of
     * these would mean riding across whatever lies between the points.
     */
    var isEstimate: Boolean = false
    var estimateReason: String? = null

    /**
     * False when the routing service could not find any route that fully honours
     * the requested avoidances and downgraded them to penalties instead. A ticked
     * "Avoid Ferries" that was silently ignored is the exact failure the avoidance
     * feature exists to prevent, so this must reach the rider.
     */
    var avoidancesHonoured: Boolean = true

    /** False when curvature data was unavailable, making the curvature figures meaningless. */
    var curvatureAvailable: Boolean = true

    /** Server-supplied explanation for either of the above. */
    var serviceWarning: String? = null

    /** Turn-by-turn navigation instructions derived from the route geometry. */
    var turnInstructions: List<TurnInstruction>? = null

    data class CurvatureMetadata(
        val totalCurvature: Double = 0.0,
        val curvaturePerKm: Double = 0.0,
        val curviestSegmentWayId: Long = 0,
        val curviestSegmentCurvature: Double = 0.0,
        val segmentsAnalyzed: Int = 0,
        val segmentsWithCurvature: Int = 0
    )
}
